ClientTests/PanelDataObserversTests.swift

```diff
@@ -36,4 +45,40 @@ class PanelDataObserversTests: XCTestCase {
 
         waitForCondition(timeout: 5) { delegate.didInvalidateCount == 3 &&  delegate.willInvalidateCount == 3 }
     }
+    
+    func testHighlightsCacheInvalidation20Min() {
+        let profile = MockProfile()
+        let observer = ActivityStreamDataObserver(profile: profile)
+        let delegate = MockDataObserverDelegate()
+        observer.delegate = delegate
+        
+        // Set to 20min since refresh
+        profile.prefs.setLong(Date.now() - (OneMinuteInMilliseconds * 20), forKey: PrefsKeys.ASLastInvalidation)
+        observer.refreshIfNeeded(forceHighlights: false, forceTopSites: false)
+        waitForCondition(timeout: 5) { delegate.highlightsRefreshCount == 1 }
+    }
+    
+    func testHighlightEmptyCache() {
+        let profile = MockProfile()
+        let observer = ActivityStreamDataObserver(profile: profile)
+        let delegate = MockDataObserverDelegate()
+        observer.delegate = delegate
+        
+        // Set to no validation key
+        profile.prefs.removeObjectForKey(PrefsKeys.ASLastInvalidation)
+        observer.refreshIfNeeded(forceHighlights: false, forceTopSites: false)
+        waitForCondition(timeout: 5) { delegate.highlightsRefreshCount == 1 }
+    }
+    
+    func testHighlightActiveCache() {
+        let profile = MockProfile()
+        let observer = ActivityStreamDataObserver(profile: profile)
+        let delegate = MockDataObserverDelegate()
+        observer.delegate = delegate
+        
+        // Set to 10min since refresh
+        profile.prefs.setLong(Date.now() - (OneMinuteInMilliseconds * 10), forKey: PrefsKeys.ASLastInvalidation)
+        observer.refreshIfNeeded(forceHighlights: false, forceTopSites: false)
+        waitForCondition(timeout: 5) { delegate.didInvalidateCount == 1 && delegate.highlightsRefreshCount == 0 }
```

Comment:
Curious if you think the timing issues you were having with the old tests could happen here again?

Reply:
Not really async but just a poorly written test. The test was originally looking for `didInvalidateCount` to be zero to verify it hadn't been called. This was also true before it was called. oops :)